python_modules/dagster/dagster/_core/definitions/resolved_asset_deps.py

```diff
@@ -187,3 +191,40 @@ def resolve_assets_def_deps(assets_defs: list[AssetsDefinition]) -> list[AssetsD
         )
 
     return resolved_assets_defs
+
+
+def resolve_stub_assets_defs(assets_defs: list[AssetsDefinition]) -> list[AssetsDefinition]:
+    defined_keys: set[AssetKey] = set()
+    deps_metadata_by_key: dict[AssetKey, list[Optional[Mapping[str, Any]]]] = defaultdict(list)
+
+    for assets_def in assets_defs:
+        for spec in assets_def.specs:
+            defined_keys.add(spec.key)
+            for dep in spec.deps:
+                deps_metadata_by_key[dep.asset_key].append(dep.metadata)
+        # treat the asset keys defined for asset checks as AssetDeps with no metadata
+        for check_spec in assets_def.check_specs:
+            deps_metadata_by_key[check_spec.asset_key].append(None)
+
+    stub_assets_defs: list[AssetsDefinition] = []
+
+    # iterate through the set of keys that are a dependency but have no definition
+    for key in deps_metadata_by_key.keys() - defined_keys:
+        # combine metadata from all dependencies and error if there are conflicts
+        metadata = {}
+        for dep_metadata in deps_metadata_by_key[key]:
+            if dep_metadata is None:
+                continue
+
+            # metadata set, but conflicts
+            if metadata and dep_metadata != metadata:
+                raise DagsterInvalidDefinitionError(
+                    f"Conflicting metadata for dependency {key.to_string()}: {metadata} != {dep_metadata}"
```

Comment:
maybe some more context in the error message about why we processing dep metadata